### ISSUE TITLE
loss nan

### ISSUE THREAD

**Issue by SSonnyboy** [fzjcdt/fgsm_uap#2]
Hello author, I'm disturbing you again. In your script, I try to observe CO in fgsm_one_uap.py with a perturbation budget of 16/255 (corresponding UAP perturbation budget is set to 32/255), where model training can normally gain losses.. But when I remove UAP (using traditional fgsm-at for adversarial training), the loss of the model becomes nan. After my observation, this seems to be caused by the output overflow of the model. How to solve this strange problem?

<img width="743" height="711" alt="Image" src="https://github.com/user-attachments/assets/a5d4899e-180a-41b4-bdaa-cec67c5fe452" />

**Comment by fzjcdt** [comment]
Hi, this is likely a gradient explosion problem. The UAP was implicitly regularizing the training, and removing it caused instability.

Try adding gradient clipping before the optimizer step. It should fix the NaN issue.


```
# train
opt.zero_grad()
loss.backward()
# Clip gradients to prevent explosion
torch.nn.utils.clip_grad_norm_(self.model.parameters(), max_norm=1)
opt.step()

```
